sam.c

```diff
@@ -2328,6 +2319,111 @@ int sam_parse1(kstring_t *s, sam_hdr_t *h, bam1_t *b)
     return -2;
 }
 
+static uint32_t read_ncigar(const char *in) {
+    uint32_t n_cigar = 0;
+    char *q = (char *)in;
+    for (; *q && *q != '\t'; ++q)
+        if (!isdigit_c(*q)) ++n_cigar;
+    if (!n_cigar) {
+        hts_log_error("No CIGAR operations");
+        return 0;
+    }
+    if (n_cigar >= 2147483647) {
+        hts_log_error("Too many CIGAR operations");
+        return 0;
+    }
+
+    return n_cigar;
+}
+
+/*! @function
+ @abstract  Parse a CIGAR string into preallocated a uint32_t array
+ @param  in      [in]  pointer to the source string
+ @param  a_cigar [out]  address of the destination uint32_t buffer
+ @return         number of processed input characters; 0 if error
+ */
+static int parse_cigar(const char *in, uint32_t *a_cigar, uint32_t n_cigar) {
+    int i, overflow = 0;
+    char *p, *q = (char *)in;
+    for (i = 0; i < n_cigar; i++) {
+        uint32_t len;
+        int op;
+        p = q;
+        len = hts_str2uint(q, &q, 28, &overflow)<<BAM_CIGAR_SHIFT;
+        if (q == p) {
+            hts_log_error("CIGAR length invalid at position %d (%s)", (int)(i+1), p);
+            return 0;
+        }
+        if (overflow) {
+            hts_log_error("CIGAR length too long at position %d (%.*s)", (int)(i+1), (int)(q-p+1), p);
+            return 0;
+        }
+        op = bam_cigar_table[(unsigned char)*q++];
+        if (op < 0) {
+            hts_log_error("Unrecognized CIGAR operator");
+            return 0;
+        }
+        a_cigar[i] = len;
+        a_cigar[i] |= op;
+    }
+
+    return q-in;
+}
+
+size_t sam_parse_cigar(const char *in, char **end, uint32_t **a_cigar, uint32_t *a_mem, uint32_t *a_len) {
+    size_t n_cigar = 0;
+    int diff;
+
+    if (!in || !end || !a_cigar || !a_mem || !a_len) {
+        hts_log_error("NULL pointer arguments");
+        return 0;
+    }
+    *end = (char *)in;
+
+    n_cigar = read_ncigar(in);
+    if (!n_cigar) return 0;
+    if (n_cigar > (*a_mem - *a_len)) {
+        uint32_t *a_tmp = realloc(*a_cigar, 1.5*(n_cigar + *a_len)*sizeof(**a_cigar));
+        if (a_tmp) {
+            *a_cigar = a_tmp;
+            *a_mem = 1.5*(n_cigar + *a_len);
+        } else {
+            hts_log_error("Memory allocation error");
+            return 0;
+        }
+    }
+
+    if (!(diff = parse_cigar(in, *a_cigar + *a_len, n_cigar))) return 0;
+    *a_len += n_cigar;
+    *end = (char *)in+diff;
+
+    return n_cigar;
+}
+
+size_t sam_parse_cigar_b(const char *in, char **end, bam1_t *b) {
```

Comment:
I don't really like the name of this function, but am struggling to think of anything I particularly like.

`sam_parse_cigar_bam` or `sam_parse_cigar2bam` maybe?

I'm not really sure it's necessary as a public function though given it could just be a tiny wrapper around `sam_parse_cigar` or `parse_cigar`?

Reply:
I knew the names would be a point of contention, so I am open to alternatives.

Reply:
Other functions with names like `…_cigar2xyz` take `n_cigar`/`const uint32_t *` as input, so using `…cigar2…` in the name wouldn't fit that pattern.

`bam_parse_cigar()` would also be a contender: there are multiple examples of `sam_abc()` is the general function, `bam_abc()` is an equivalent limited to BAM in some sense (here, in that it's for `bam1_t`, like other `bam_` functions as discussed in https://github.com/samtools/htslib/pull/1159#issuecomment-715280626).

However there is an extent to which wanting this as a public function really shows a deficiency in the recently-added `bam_set1()`: perhaps that should be able to take either a `size_t n_cigar, const uint32_t *cigar ` or a `const char *cigar_string`. It does after all take other unparsed string parameters e.g. `seq`, `qual`…

Reply:
I think there is scope for wanting both text construction and binary construction.

Take CRAM for example.  It's constructing the cigar data from the feature codes.  So it already has data like code = 'D', length = 10, etc.  It would be ugly to have to construct "10D20M" etc in a string only to parse it back again.  Also consider cases where we want to do manipulation to an existing record - eg turning soft-clips into hard-clips.  (The sequence is problematic given it's annoying nibblyness, but we have to live with that so it's inevitable we end up in 1-byte sequence world or have a two code paths for len%2. )  We already have int based cigars, so again having to go into and out of string form would be poor.  Other sources of data may differ though and start from strings.

If we only have one interface, the binary form is better as you can first call one function to parse and then pass those parameters into `bam_set1`.